chromium_src/chrome/browser/profiles/profile_window.cc

```diff
@@ -43,6 +46,32 @@ void OpenBrowserWindowForTorProfile(ProfileManager::CreateCallback callback,
       profile->GetOffTheRecordProfile(), status);
 }
 
+void OnTorRegularProfileCreated(ProfileManager::CreateCallback callback,
+                                bool always_create,
+                                bool is_new_profile,
+                                bool unblock_extensions,
+                                Profile* profile,
+                                Profile::CreateStatus status) {
+  // We need to postpone the timing of creating the off-the-record Tor profile
+  // and let the regular Tor profile finish ProfileManager::DoFinalInit first.
+  // So we pass the regular Tor profile here and access the off-the-record Tor
+  // profile later when this task is executed.
+  //
+  // It is because an existing ProfileObserver might only start to observe
+  // this regular profile in OnProfileAdded which is happened in DoFinalInit.
+  // For example, ChromeProcessManagerDelegate has this behavior and we will
+  // miss the OnOffTheRecordProfileCreated event if we create the
+  // off-the-record Tor profile before OnProfileAdded of the regular Tor
+  // profile is called. This would lead us to not destroying the background
+  // hosts in time before the off-the-record Tor profile is destroyed and hit
+  // DCHECK in ProfileDestroyer because a render process host wasn't destroyed
+  // before the off-the-record profile is destroyed.
+  base::PostTask(FROM_HERE, {content::BrowserThread::UI},
```

Comment:
I think the problem is in `ProfileManager::OnProfileCreated` because the callback is trigged here
```c++
1127   // Invoke CREATED callback for normal profiles.
1128   bool go_off_the_record = ShouldGoOffTheRecord(profile);
1129   if (success && !go_off_the_record)
1130     RunCallbacks(callbacks, profile, Profile::CREATE_STATUS_CREATED);
```
but when that happens `info->created` is false which means `DoFinalInit` is not called (it is called right after this block).
Maybe we can check something like `!IsTorProfile` here to prevent callback from being called prematurely

Reply:
Personally I would prefer the PostTask solution over adding a patch here. It seems more compatible with upstream to me by following all their flow in this function, and also mainly because of avoiding patching.

Reply:
It might potentially break again in future chromium upgrade of course, but that should be caught by our auto test like this time hopefully.

Reply:
Another possible solution I could think of is adding a class similar to https://source.chromium.org/chromium/chromium/src/+/master:chrome/browser/profiles/profile_window.h;l=137?originalUrl=https:%2F%2Fcs.chromium.org%2F, we'll subclass `ProfileObserver` instead in this case.
When the profile creation callback is run, we instantiate this observer class, observe the created profile, and pass all the arguments we need for `OpenBrowserWindowForProfile` into it.
Then execute  `OpenBrowserWindowForProfile` when `OnProfileAdded` is called (observed).